### ISSUE TITLE
Unknown type with `dict` as argument to generic function

### ISSUE THREAD

**Issue by robert-bn** [microsoft/pyright#10620]
With `reportUnknownVariableType` set to "error", the following code produces an error in pyright 1.1.402

```python
from collections.abc import Callable

def field[T](default_factory: Callable[[],T]) -> T: ...

example: dict[str, str] = field(default_factory=dict)
```
the error is:
>error: Type of "example" is partially unknown
>    Type of "example" is "dict[Unknown, Unknown]" (reportUnknownVariableType)

This affects the reasonably common pattern of using `dict` as a default factory in dataclass fields. 

I'm not completely sure if this is a bug or intended behaviour, but I would have expected pyright to be able to handle this case, as pyright successfully determines `dict` is an instance of `Callable[[], dict[str,str]]`. Inserting a function call in between assignment to `example` to ensure pyright throws away the more specific type information of `dict` and see it only as `Callable[[], dict[str,str]]` results in no error in the assignment:

```python
from collections.abc import Callable

def field[T](default_factory: Callable[[],T]) -> T: ...

def f(mk_dict: Callable[[], dict[str,str]]):
    _example: dict[str, str] = field(default_factory=mk_dict)

f(dict)
```

**Comment by Daraan** [comment]
I believe this is somewhat behavior here. `field` returns `dict[Unknown, Unknown]` here, afaik a custom annotation should not silence the error. I encountered this once too, you can solve it by using:
`example: dict[str, str] = field(default_factory=dict[str, str])`  // depends on your Python version
for other python versions or if you want to avoid the creation of the runtime GenericAlias use:
`example: dict[str, str] = field(default_factory=dict[str, str] if TYPE_CHECKING else dict)`

**Comment by erictraut** [comment]
Duplicate of https://github.com/microsoft/pyright/issues/10283.
